### PR TITLE
feature(provision/aws): upfront AZ filtering for capacity awareness

### PR DESCRIPTION
**Pull request by dimakr**
Phases 1-2 implementation of the plan: [docs/plans/infrastructure/aws-capacity-az-fallback.md](https://github.com/scylladb/scylla-cluster-tests/blob/da3ffd7d51b90b15f10b922b029cc8c6f92dc41b/docs/plans/infrastructure/aws-capacity-az-fallback.md)

Tests frequently fail at provision time when the configured AZ does not offer the requested instance type.
The change centralize AZ discovery and capacity-error classification. Adds AZResolver that runs before provisioning in both legacy and modern paths to drop unsupported AZs (replacing them with valid alternatives in the same region). Covers existing role types plus dynamically-added ones (instance_type_db_target, nemesis_grow_shrink_instance_type).
Raises exception before any RunInstances call, when no AZ supports the full type set.

Gated by pre_filter_unavailable_availability_zones SCT config parameter (default true).

Refs: https://scylladb.atlassian.net/browse/SCT-295

### Testing

- [x] [invalid AZ 'f' in us-east-1 is replaced with valid alternative](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dimakr/job/pr-provision-test/267/)
AZResolver detects that the configured AZ does not offer all required instance types, replaces with a valid alternative in the same region, and provisioning proceeds in the new AZ.

Some logs from the run:
AZResolver fired with replacement
```
❯ grep "AZResolver:" sct-55aead5f.log
< t:2026-05-08 23:33:54,455 f:az_resolver.py  l:147  c:sdcm.provision.aws.az_resolver p:WARNING > AZResolver: availability_zone 'f' does not support all required instance types ['i8g.2xlarge', 'c6i.xlarge', 't3.large'] in regions ['us-east-1']; replacing with 'a'
```
Prefered AZ from AwsRegion:
```
❯ grep "Preferred availability zones\|Availability zones in us-east-1 supporting" sct-55aead5f.log
< t:2026-05-08 23:33:54,453 f:aws_region.py   l:123  c:sdcm.utils.aws_region p:WARNING > Preferred availability zones ['us-east-1f'] do not support all required instance types ['i8g.2xlarge', 'c6i.xlarge', 't3.large']
< t:2026-05-08 23:33:54,453 f:aws_region.py   l:131  c:sdcm.utils.aws_region p:INFO  > Availability zones in us-east-1 supporting ['i8g.2xlarge', 'c6i.xlarge', 't3.large']: ['us-east-1a', 'us-east-1b', 'us-east-1c', 'us-east-1d']
```
EC2 activities moved to replacement AZ - no `us-east-1f` log entries after the resolver handled the replacement:
```
❯ grep "us-east-1f" sct-55aead5f.log
< t:2026-05-08 23:33:54,453 f:aws_region.py   l:123  c:sdcm.utils.aws_region p:WARNING > Preferred availability zones ['us-east-1f'] do not support all required instance types ['i8g.2xlarge', 'c6i.xlarge', 't3.large']
```
Subnet lookup in the replacement AZ
```
❯ grep -E "AvailabilityZone.: .us-east-1a.|SCT-2-subnet-us-east-1a" sct-55aead5f.log | head
< t:2026-05-08 23:33:55,191 f:aws_region.py   l:153  c:sdcm.utils.aws_region p:DEBUG > Found Subnets: {'Subnets': [{'AvailabilityZoneId': 'use1-az1', 'MapCustomerOwnedIpOnLaunch': False, 'OwnerId': '797456418907', 'AssignIpv6AddressOnCreation': True, 'Ipv6CidrBlockAssociationSet': [{'AssociationId': 'subnet-cidr-assoc-018122b708b8b5919', 'Ipv6CidrBlock': '2600:1f18:7f4:3c00::/64', 'Ipv6CidrBlockState': {'State': 'associated'}, 'Ipv6AddressAttribute': 'public', 'IpSource': 'amazon'}], 'Tags': [{'Key': 'Name', 'Value': 'SCT-2-subnet-us-east-1a'}], 'SubnetArn': 'arn:aws:ec2:us-east-1:797456418907:subnet/subnet-0a09ba4421ec6aaa8', 'EnableDns64': False, 'Ipv6Native': False, 'PrivateDnsNameOptionsOnLaunch': {'HostnameType': 'ip-name', 'EnableResourceNameDnsARecord': False, 'EnableResourceNameDnsAAAARecord': False}, 'BlockPublicAccessStates': {'InternetGatewayBlockMode': 'off'}, 'SubnetId': 'subnet-0a09ba4421ec6aaa8', 'State': 'available', 'VpcId': 'vpc-0935da7392dc9d4d8', 'CidrBlock': '10.12.0.0/22', 'AvailableIpAddressCount': 986, 'AvailabilityZone': 'us-east-1a', 'DefaultForAz': False, 'MapPublicIpOnLaunch': True}], 'ResponseMetadata': {'RequestId': '0a18d4c5-219e-48eb-b846-5bd0eb99170a', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amzn-requestid': '0a18d4c5-219e-48eb-b846-5bd0eb99170a', 'cache-control': 'no-cache, no-store', 'strict-transport-security': 'max-age=31536000; includeSubDomains', 'content-type': 'text/xml;charset=UTF-8', 'content-length': '1675', 'date': 'Fri, 08 May 2026 23:33:55 GMT', 'server': 'AmazonEC2'}, 'RetryAttempts': 0}}
```
- [x] [no AZ supports all specified types; fail fast before any instances are run](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dimakr/job/pr-provision-test/268/)
The exception is raised during provisioning step (as expected):
```
12:50:14  sdcm.provision.aws.az_resolver.NoValidAvailabilityZoneError: No availability zone supports all required instance types ['dl1.24xlarge', 'hpc7g.16xlarge', 't3.large'] across regions ['us-east-1']; cannot proceed with provisioning.
```

### PR pre-checks (self review)

- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

### CODE REVIEW

File: sdcm/provision/aws/capacity_reservation.py [hunk 1]
```diff
@@ -87,28 +88,13 @@ def get_cr_from_aws(cls, params, force_fetch=False) -> None:
...
-            LocationType="availability-zone",
-            Filters=[
-                {"Name": "instance-type", "Values": list(instance_types)},
-            ],
+    def _get_supported_availability_zones(ec2, instance_types: List[str], initial_az: str) -> List[str]:  # noqa: ARG004
+        # ec2 client in signature is for backward compatibility (AwsRegion creates its own boto3 client)
+        aws_region = AwsRegion(region_name=initial_az[:-1])
+        return aws_region.get_common_availability_zones(
+            instance_types=list(instance_types),
+            preferred_azs=[initial_az],
```
[the remark this answers is not in the document]
Comment (dimakr) [reply sdcm/provision/aws/capacity_reservation.py]: Fixed by passing region and availability_zone separately

File: sdcm/provision/aws/az_resolver.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            )
+            self._params["availability_zone"] = new_value
+        else:
+            LOGGER.info("AZResolver: availability_zone '%s' already valid for region %s", original_value, region_name)
+
+    def _region_name(self) -> str:
+        if regions := getattr(self._params, "region_names", None):
+            return regions[0]
+        parts = (self._params.get("region_name") or "").split()
+        return parts[0] if parts else ""
```
[the remark this answers is not in the document]
Comment (dimakr) [reply sdcm/provision/aws/az_resolver.py]: Fixed. resolve() now intersects supported AZ letters across all region_names

File: unit_tests/unit/provisioner/test_az_resolver.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            "availability_zone": "a",
+            "instance_type_db": "i4i.large",
+            "instance_type_loader": "t3.small",
+            "instance_type_monitor": "t3.small",
+            "pre_filter_unavailable_availability_zones": True,
+            **overrides,
+        }
+    )
+    params.region_names = params["region_name"].split()
+    return params
```
[the remark this answers is not in the document]
Comment (dimakr) [reply unit_tests/unit/provisioner/test_az_resolver.py:44]: Added test_resolve_multi_region_intersects_supported_az_letters (happy path) and test_resolve_multi_region_raises_when_no_common_letter (no-overlap negative)

File: sdcm/provision/aws/capacity_errors.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    "InsufficientCapacity",
+]
+
+
+def is_capacity_error(exception: BaseException) -> bool:
+    """Return True if `exception` represents AWS capacity shortage."""
+    if not isinstance(exception, ClientError):
+        return False
+    error_code = exception.response.get("Error", {}).get("Code", "")
+    return error_code in CAPACITY_ERROR_KEYWORDS or any(kw in str(exception) for kw in CAPACITY_ERROR_KEYWORDS)
```
[the remark this answers is not in the document]
Comment (dimakr) [reply sdcm/provision/aws/capacity_errors.py]: Fixed: dropped the broad str(exc) substring scan. Now matching Error.Code exactly

File: sdcm/provision/aws/az_resolver.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+from sdcm.utils.aws_region import AwsRegion
+
+LOGGER = logging.getLogger(__name__)
+
+_INSTANCE_TYPE_PARAM_KEYS: tuple[str, ...] = (
+    "instance_type_db",
+    "instance_type_loader",
+    "instance_type_monitor",
+    "zero_token_instance_type_db",
```
[the remark this answers is not in the document]
Comment (dimakr) [reply sdcm/provision/aws/az_resolver.py]: Added instance_type_vector_store

File: sdcm/provision/aws/az_resolver.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    """Resolve `availability_zone` config to AZs supporting all required instance types."""
+
+    def __init__(self, params):
+        self._params = params
+
+    def required_instance_types(self) -> list[str]:
+        """Get the deduplicated list of instance types a test will launch."""
+        seen = []
+        for key in _INSTANCE_TYPE_PARAM_KEYS:
+            try:
```
[the remark this answers is not in the document]
Comment (dimakr) [reply sdcm/provision/aws/az_resolver.py]: Fixed. Each instance_type_* key now maps to a node-count / feature-flag gate

File: defaults/test_default.yaml [hunk 7]
```diff
@@ -213,6 +213,8 @@ bare_loaders: false
 
 aws_fallback_to_next_availability_zone: false
 
+pre_filter_unavailable_availability_zones: true
```
Comment (fruch) [inline defaults/test_default.yaml:216]: @dimakr is there a reason we don't want to check this ? for the test that has those configuration ? or that the way to mark them ?

**Review by fruch — COMMENTED** [review]
# PR Review Summary

**PR**: `feature(provision/aws): upfront AZ filtering for capacity awareness`
**Files**: 12 changed (+636, -22)

---

## Critical Issues (3)

### 1. Silent `except Exception` blocks in `az_resolver.py:93-100`

Two bare `except Exception` catches with `noqa: BLE001` silently swallow errors and default to potentially wrong values (`value = None`, `gate_open = True`). If a gate function has a bug, AZ filtering silently includes/excludes wrong instance types, causing confusing provisioning failures later with no trail back.

```python
try:
    value = self._params.get(key)
except Exception:  # noqa: BLE001
    value = None
# ...
try:
    gate_open = gate(self._params)
except Exception:  # noqa: BLE001
    gate_open = True
```

**Fix**: Remove catch-alls or narrow to specific expected exceptions (e.g. `KeyError`) and add `LOGGER.debug()`.

### 2. `capacity_errors.py` is dead code

Not imported by any production code — only by its test file. This module is scaffolding for Phase 2.

**Fix**: Either defer to a later PR or add a module-level docstring: `# Phase 2 — not yet consumed; see docs/plans/infrastructure/aws-capacity-az-fallback.md`.

### 3. `dedicated_host.py` has duplicate `_get_supported_availability_zones` NOT refactored

`capacity_reservation.py` was refactored to delegate to `AwsRegion.get_common_availability_zones()`, but `dedicated_host.py` has an identical copy of the OLD implementation still using raw boto3 calls. The plan document explicitly calls out eliminating this duplication.

---

## Important Issues (3)

### 4. `_node_count_positive` has no direct tests

This is the gating logic that determines which instance types constrain AZ selection. All branches (`bool`, `int`, `float`, `list`, `str`, `None`, negative values) should be directly tested with `@pytest.mark.parametrize`. A bug here (e.g., treating `"0"` as truthy) would cause unnecessary AZ filtering failures in production.

### 5. `CAPACITY_ERROR_KEYWORDS` naming implies substring matching but code does exact set intersection

`"Unsupported"` won't match `"UnsupportedOperation"` — the test at line 36 passes by accident of exact-match semantics. If intent is exact error codes, rename to `CAPACITY_ERROR_CODES`. The test comment _"'Unsupported' as a substring of a different code must not match"_ is misleading since no substring matching exists.

### 6. Dual `AZResolver.resolve()` call sites

Called from both `tester.py:1905` (legacy path) and `layout.py:32` (modern path). Needs a comment clarifying these are mutually exclusive paths, or a guard to prevent double invocation + redundant AWS API calls.

---

## Suggestions

- Add test for `availability_zone=""` / `None` in `resolve()` — behavior of auto-selecting all valid AZs should be explicit
- Add test for `get_common_availability_zones([])` — the early return path
- Test `instance_type_db_target=None` and `nemesis_grow_shrink_instance_type=""` explicitly (these use `_always` gate)
- Verify existing `test_capacity_reservation.py` tests still pass — the refactored method now creates a new `AwsRegion` instance internally instead of using the `ec2` client, which may bypass existing mocks

---

## Strengths

- Clean architecture: `AwsRegion.get_common_availability_zones()` centralizes AZ discovery well
- Proper pytest style throughout, good fixture design (especially `mock_multi_region`)
- Config default present in `defaults/test_default.yaml` ✅
- Import conventions followed ✅
- `NoValidAvailabilityZoneError` gives clear, actionable error messages
- Multi-region intersection logic is correct and well-tested
- Override safety: no subclass overrides affected by the `_get_supported_availability_zones` signature change

File: defaults/test_default.yaml (hunk 7 above)
Comment (dimakr) [reply defaults/test_default.yaml:216]: The flag and true default are per the plan (docs/plans/infrastructure/aws-capacity-az-fallback.md, Phase 2).
The pre-filter is a no-op when all AZs support the required types, so enabling it by default normally has no costs, and only fires when a selected AZ is unusable for the instances selected for a test (we can disable this switch for tests that have justified reason for that).

**Comment by dimakr** [comment]
Addressed comments C1, C3, I4, I5, S1, S3 from the round of review https://github.com/scylladb/scylla-cluster-tests/pull/14561#pullrequestreview-4259410540 above.
For the remaining commencs:
 - C2: capacity_errors.py is dead code: this is 'by design' - as per the implementation plan the consumers of this module will land in Phases 3 and 4. So for now it;s a prep work for future phases.
 - I6: dual AZResolver.resolve() call sites: modern (SCTProvisionAWSLayout.provision) and legacy (tester.get_cluster_aws) provisioning paths are mutually exclusive - only one runs per test (controlled by config). AZResolver.resolve() is also idempotent, as the  second call on already-resolved params produces the same result.
 - S2: get_common_availability_zones([]) test - AZResolver.resolve() already exists early with "No instance types declared..." message before reaching get_common_availability_zones, so such a test would add a code path that is not exercised.
 @fruch

**Comment by fruch** [comment]
We should test also multiple AZs case (or to the minimum add those kind of unittests)

**Comment by dimakr** [comment]
Multi-AZ unit tests (there are few cases) exist now in `unit_tests/unit/provisioner/test_az_resolver.py`.
 
Also executed the tests similar to how single AZ case was tested (described in PR description):
- [no AZ supports all required types; fail fast before any instances are run](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dimakr/job/pr-provision-test/275/)

The parameters used:
```
SCT_REGION_NAME=us-east-1
SCT_AVAILABILITY_ZONE=a,b,c
SCT_INSTANCE_TYPE_DB=dl1.24xlarge        # only us-east-1b
SCT_INSTANCE_TYPE_LOADER=hpc7g.16xlarge  # only us-east-1d
```
failed fast with the `21:59:45  sdcm.provision.aws.az_resolver.NoValidAvailabilityZoneError: No availability zone supports all required instance types ['dl1.24xlarge', 'hpc7g.16xlarge', 't3.large'] across regions ['us-east-1']; cannot proceed with provisioning.` on Provisioning step

- [Invalid AZ from set of multi-AZs was replaced with valid alternative, others are kept](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dimakr/job/pr-provision-test/277/)

Some log entries that indicate that replacement was performed:
Logging of which AZs are to be rejected, which are available for selection:
```
❯ grep -E "Preferred availability zones|Availability zones in .* supporting" sct-5e6fa305.log
< t:2026-05-11 20:50:58,160 f:aws_region.py   l:123  c:sdcm.utils.aws_region p:WARNING > Preferred availability zones ['us-east-1e'] do not support all required instance types ['i4i.4xlarge', 'c6i.xlarge', 't3.large']
< t:2026-05-11 20:50:58,160 f:aws_region.py   l:131  c:sdcm.utils.aws_region p:INFO  > Availability zones in us-east-1 supporting ['i4i.4xlarge', 'c6i.xlarge', 't3.large']: ['us-east-1a', 'us-east-1b', 'us-east-1c', 'us-east-1d', 'us-east-1f']
```
AZ replacement happened
```
❯ grep "AZResolver: availability_zone" sct-5e6fa305.log
< t:2026-05-11 20:50:58,163 f:az_resolver.py  l:138  c:sdcm.provision.aws.az_resolver p:WARNING > AZResolver: availability_zone 'a,b,e' does not support all required instance types ['i4i.4xlarge', 'c6i.xlarge', 't3.large'] in regions ['us-east-1']; replacing with 'a,b,c'
```
Indication which AZs the nodes landed on (invalid `e` AZ is not there):
```
❯ grep -oE "'AvailabilityZone': 'us-east-1[a-f]'" sct-5e6fa305.log
'AvailabilityZone': 'us-east-1a'
'AvailabilityZone': 'us-east-1b'
'AvailabilityZone': 'us-east-1c'
```
@fruch

**Review by fruch — APPROVED** [review]
LGTM

let's move to the next stage
